### DISCUSSION TITLE
Reacher.calc_state() doesn't work as expected

### DISCUSSION THREAD

**Discussion by ven-kyoshiro** [bulletphysics/bullet3#3659] [General]
## State space is overlapping  and not natural

https://github.com/bulletphysics/bullet3/blob/c4300dc2999b37ac0f689a54df422434a6cf5ddd/examples/pybullet/gym/pybullet_envs/robot_manipulators.py#L29

I think this row should use 
> self.central_joint.current_position()

Because the domain of (sin,cos) is overlapped.
For example,
```
env = ReacherBulletEnv()
env.reset()

# for all a
a = 0.1
env.robot.central_joint.reset_current_position(a, 0)
s1 = env.robot.calc_state()

env.robot.central_joint.reset_current_position(a+np.pi, 0)
s2 = env.robot.calc_state()
s1[4:6] ==s2[4:6]
# array([ True,  True])
```

**Comment by manuelbre** [comment]
I came here to post about the same topic and found this open issue.

I agree with the statement that this can be solved by using `self.central_joint.current_position()` . 
In addition, I believe that it would make sense to also use `self.central_joint.current_position()` for `gamma` although I believe that it is correct as is, but using the `self.central_joint.current_position()`  makes the state interpretable without having to search for the upper and lower limit of the elbow joint.

I would propose to change https://github.com/bulletphysics/bullet3/blob/c4300dc2999b37ac0f689a54df422434a6cf5ddd/examples/pybullet/gym/pybullet_envs/robot_manipulators.py#L29-L44 to:

````
...
theta, self.theta_dot = self.central_joint.current_position()
self.gamma, self.gamma_dot = self.elbow_joint.current_position()
...
````

Connected to this issue is that it would also make sense to use `cos(gamma), sin(gamma)` instead of simply using `gamma` as the state. Let me know if I should post a second issue about this.
